Adding GEP 957 for port matching support.

cxhiano opened the pull request:
**What type of PR is this?**
/kind gep
/kind api-change

**What this PR does / why we need it**:
Proposal to support destination port matching.

**Which issue(s) this PR fixes**:
Fixes #957 

**Does this PR introduce a user-facing change?**:
```release-note
Add a new field `Port` to `ParentRef` for destination port matching
```

robscott commented:
/ok-to-test
/hold for consensus

robscott commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,77 @@
+
+## Goals
+
+* Support destination port matching based on port number.
```
In the community meeting yesterday I think there was a bit of confusion around what destination port meant, this might be a bit clearer, but not sure:
```suggestion
* Support port matching in routes based on the destination port number of the request.
```

robscott commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,77 @@
+
+## TLDR
+
+Support destination port matching.
```
```suggestion
Add a new `port` field to ParentRef to support port matching in Routes.
```

robscott commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,77 @@
+
+## Alternatives
+### 1. Use SectionName in ParentRef for port matching
+Port matching can be supported if SectionName accepts port numbers in addition
+to listener names. This approach results in a less explicit API. It can also
+complicate the validation on SectionName.
```
Maybe add a bit more detail here such as an example of how it would complicate SectionName validation and/or how this would provide a poor user experience.

robscott commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,77 @@
+example, some users may only want to expose their application on ports 80 or 443
+on the Gateway.
+
+## API
```
Probably need a section in the GEP and/or godoc below describing the interaction between SectionName and Port. I'm not actually sure what the best approach is though.

We could say they were mutually exclusive, but that is likely unnecessarily restrictive. For ingress, one could say that they only wanted to attach to Gateway listener named foo if it was also listening on port 80. I'm assuming that wouldn't be super common, but it does seem fairly easy to interpret. For mesh, section name could represent some other implementation specific attribute such as a network name. In that case, it seems like specifying both network/sectionName and port could be quite helpful.

Given the potential usefulness, maybe we should suggest that these fields are complementary?

cxhiano replied:
Yeah, I don't have a strong opinion on this. Agreed that it may be better to make these fields complementary given that there is a reasonable interpretation on the meaning when both fields are present.

robscott reviewed (commented):
Thanks for your work on this @cxhiano!

cxhiano commented:
Thanks for the comments! PR updated. PTAL!

robscott commented:
@hbagdi had a good comment in Slack asking if port would really be sufficient in `parentRef` or if we'd also want to add more fields in the future. Although it's hard to predict the future, I think it's at least possible that more fields could be added here. When I originally proposed `sectionName`, I remember at least 2 other alternatives being discussed:

1. A label selector that could enable selecting multiple parent resources. Although it's possible we may add this in the future, I think all the selectors in v1alpha1 helped show just how complex selectors can become. I personally don't think this is likely.
2. A jsonpath field that would enable matching arbitrary values in any resource. This would also provide a great deal of flexibility (and complexity), but I'm not currently aware of any use cases for it. I'd personally be hesitant to add in this kind of complexity.

It is possible that there are other potential extension points I'm missing here, but I can't think of any obvious ones. Even if we did want to add a field in the future, I don't think there's anything in this proposal that would prevent that from happening. With that said, very interested in what others think.

jpeach commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,82 @@
+cases. For example, it's less inituitive to create a route with
+`ParentRef.SectionName` set to `443` to express `route all traffic whose
+destination port is 443 to ...`. It also complicates the validation on
+SectionName in order to differentiate between a listener name and a port number.
```
IIUC the only reason that section name isn't sufficient is where the parent is some resource that is not a Gateway. It would be good to call that out here because it's pretty important IMHO.

cxhiano replied:
Rephrased this section. Please let me know if it is more clear now.

robscott commented:
@cxhiano we discussed this PR in more detail at today's community meeting. There was a lot of good feedback, I'll try to summarize it here, but of course anyone should feel free to add anything I miss.

1. It would be useful to have some more concrete examples of how this would be used for both Gateway and Mesh. This would help explain the rationale for the GEP and likely provide a better basis for the rest of the discussion. 
2. There's some concern that users using port number will more tightly couple the relationship between Gateways and Routes. When app devs attach Routes with a port number, it means that Gateway admins no longer have the flexibility to switch ports on the Gateway without also updating the Routes. I think this will be a reasonable tradeoff to make for some, but we should add some guidance to the GEP, and eventually documentation around when users would want to use one approach vs the other.
3. There was also some concern about this providing too many ways to do the same thing. I think that could largely be resolved by addressing the first 2 items (clarifying the rationale for adding this capability + clearly defining when each field should be used).

cxhiano commented:
Thanks for the feedback! Updated the API section to provide a recommendation on when `port` should be used. Also added two examples on how it can be used. (One for Gateway, the other for Mesh)

hbagdi commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,140 @@
+
+## Non-Goals
+
+* Support port matching based on named port.
```
```suggestion
* Support port matching based on port number.
```
Named port is ambiguous in this context (port names in k8s, section names, etc)

hbagdi commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,140 @@
+## Introduction
+
+Port matching is a common service mesh use case where traffic policies/rules
+need to be applied to traffic to certain destination ports. For ingress, while
+the API today already supports attaching a route to a specific listener, it may
+be useful to support attaching routes to listener(s) on a specified port. For
+example, some users may only want to expose their application on ports 80 or 443
+on the Gateway.
```
> be useful to support attaching routes to listener(s) on a specified port

Why might that be useful?
Please elaborate on that for the purpose of this GEP.

80/443 is not the best example here probably. Readers are not imaginative so picking a better example will bring more clarity.

cxhiano replied:
Updated this section and also the field comment in API section to be more prescriptive. 

Admittedly while there are clear, concrete mesh use cases, it's only potentially useful for ingress(@robscott please chime in if I am missing a concrete ingress use case here). I am hoping that this GEP is a reasonable middle ground which enhances this API to be more inclusive towards use cases outside of ingress(such as mesh) while not making it more confusing/complicated for the existing ingress support.

hbagdi commented:
My concern is that this has the potential to complicate Ingress use-cases.
Route authors, upon seeing this, could become more defensive than necessary and put in port numbers.
Can we be more prescriptive about specific cases where this must be used?

pleshakov commented:
If the proposed `port` field can potentially complicate the API for the Gateway kind, perhaps it makes sense to consider adding extension options to the `parentRef`, so that third-party kinds can leverage them?

For example:
```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: HTTPRoute
metadata:
  name: myroute
spec:
  parentRefs:
  - name: my-mesh
    group: example.io
    kind: Mesh
    options:
      port: "8080"
  hostnames:
  - example.com
. . .
```
where `options` is similar to https://github.com/kubernetes-sigs/gateway-api/blob/60e1838a179b4960ed5d0f825bf8ebfa189b8580/apis/v1alpha1/gateway_types.go#L360

robscott commented:
> My concern is that this has the potential to complicate Ingress use-cases. Route authors, upon seeing this, could become more defensive than necessary and put in port numbers. Can we be more prescriptive about specific cases where this must be used?

I think for Ingress this is adding one more optional component to a handshake. While it could lead to more defensive configuration from Route owners, I don't think that would always be a bad thing. I think it would be highly disruptive to app/route owners if a Gateway admin changed the ports that are being listened on. It seems unlikely that this would be a frequently changed attribute of a Gateway.

I think it's more likely that a Gateway admin would want to change a listener name than the listener port. For example, maybe a listener was initially focused on just app A, but they want to expand that listener to cover app A and B. Or maybe they want to split a shared listener into different ones? 

If a Route owner can attach to example.com on port 80, that may be more resilient (and easier to understand) than attaching to a relatively abstract section name. To be clear, I think section name will still be a useful mechanism for many orgs, I'm just starting to think that may become the advanced case, not the default.

robscott commented:
>If the proposed port field can potentially complicate the API for the Gateway kind, perhaps it makes sense to consider adding extension options to the parentRef, so that third-party kinds can leverage them?

Although we have that for TLS options, we've generally tried to avoid open ended configs like that. I think wherever we have a widely understand concept, we should try to surface that to a top level field if possible. On the other hand, if there are some use cases that rely on implementation specific concepts and terminology, it may be worth exploring a more open ended `options` field. @pleshakov do you have any use cases in mind beyond port?

pleshakov commented:
@robscott 
> , we've generally tried to avoid open ended configs like that. I think wherever we have a widely understand concept, we should try to surface that to a top level field if possible. 

That makes sense.

> On the other hand, if there are some use cases that rely on implementation specific concepts and terminology, it may be worth exploring a more open ended options field. @pleshakov do you have any use cases in mind beyond port?

Unfortunately, I don't have specific examples beyond the port. But I'm not too familiar with the service mesh use cases.

hbagdi commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,140 @@
+# GEP-957: Destination Port Matching
+
+* Issue: [#957](https://github.com/kubernetes-sigs/gateway-api/issues/957)
+* Status: Provisional
```
Implementable?

hbagdi commented on site-src/geps/gep-957.md:
````diff
@@ -0,0 +1,140 @@
+```go
+type ParentRef struct {
+  ...
+  // Port is the network port this resource targets. It can be interpreted
````
nit:
> Port is the network port this resource targets

What is `this`? Very ambiguous.

cxhiano replied:
Updated to "this Route"

hbagdi commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,140 @@
+  // differently based on the type of parent resource:
+  //
+  // Gateway: All listeners listening on the specified port that also support
+  // this kind of Route. It's only recommended to set `Port` when the networking
```
```suggestion
  // this kind of Route (and select this Route). It's only recommended to set `Port` when the networking
```

hbagdi commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,140 @@
+  // When both Port and SectionName are specified, the name and port of the
+  // selected listener must match both specified values.
+  //
+  // Implementations MAY choose to support other parent resources.
+  // If that is the case, they MUST clearly document how Port is
+  // interpreted.
```
```suggestion
  // Implementations supporting other types of parent resources
  // MUST clearly document how/if Port is interpreted.
```

hbagdi commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,140 @@
+  // If that is the case, they MUST clearly document how Port is
+  // interpreted.
+  //
+  // When both Port and SectionName are unspecified, this will reference the
+  // entire resource.
```
Recommend leaving this out. Port field is a more specific matcher and should refrain from general behavior guidance.

hbagdi commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,140 @@
+  // When both Port and SectionName are unspecified, this will reference the
+  // entire resource.
+  //
+  // For the purpose of status, an attachment is considered successful as
+  // long as the parent resource accepts it partially. For example, Gateway
```
```suggestion
  // For the purpose of status, an attachment is considered successful if
  // at least one listener accepts the Route. For example, Gateway
```

cxhiano replied:
I wonder if we can phrase this in a way that is agnostic to the type of parent resource here while the example following this general statement is specifically about Gateway/Listener?

hbagdi commented:
> I think for Ingress this is adding one more optional component to a handshake. While it could lead to more defensive configuration from Route owners, I don't think that would always be a bad thing. I think it would be highly disruptive to app/route owners if a Gateway admin changed the ports that are being listened on. It seems unlikely that this would be a frequently changed attribute of a Gateway.

It really comes down to abstractions that should be exposed to a human. I think abstracting away port at this layer is really hard and this GEP furthers this fact.
Like you say it is not a bad thing, it just increases the number of things a Gateway and a Route owner must cross-check.

I still think we should strongly advise users to not use this field for Ingress-like use-cases to unnecessarily avoid complicating the simpler use-cases. If we can add that bit, I'm good with this GEP.

cxhiano commented:
Thanks for the review!

> I still think we should strongly advise users to not use this field for Ingress-like use-cases to unnecessarily avoid complicating the simpler use-cases. If we can add that bit, I'm good with this GEP.

I updated the API documentation to say "It's not recommended to use `Port` unless...". Please let me know if that is sufficient.

hbagdi commented on site-src/geps/gep-957.md:
```diff
@@ -0,0 +1,138 @@
+  // differently based on the type of parent resource:
+  //
+  // Gateway: All listeners listening on the specified port that also support
+  // this kind of Route(and select this Route). It's not recommended to set
+  // `Port` unless the networking behaviors specified in a Route should only
+  // apply to a specific port as opposed to listener(s) whose port(s) may
+  // get switched by Gateway admins.
```
```suggestion
  // this kind of Route(and select this Route). It's not recommended to set
  // `Port` unless the networking behaviors specified in a Route must
  // apply to a specific port as opposed to a listener(s) whose port may
  // be changed.
```

robscott commented:
Thanks @cxhiano!

/lgtm

robscott commented:
/hold cancel
